docker/docker_transport.go

```diff
@@ -0,0 +1,120 @@
+	return ParseReference(reference)
+}
+
+// dockerReference is an ImageReference for Docker images.
+type dockerReference struct {
+	ref reference.Named
+	tag string
+}
+
+// ParseReference converts a string, which should not start with the ImageTransport.Name prefix, into an Docker ImageReference.
+func ParseReference(reference string) (types.ImageReference, error) {
+	return parseReference(reference)
+}
+
+// ParseReference converts a string, which should not start with the ImageTransport.Name prefix, into an dockerReference.
+func parseReference(refString string) (dockerReference, error) {
+	if !strings.HasPrefix(refString, "//") {
+		return dockerReference{}, fmt.Errorf("docker: image reference %s does not start with //", refString)
+	}
+	ref, err := reference.ParseNamed(strings.TrimPrefix(refString, "//"))
+	if err != nil {
+		return dockerReference{}, err
+	}
+	ref = reference.WithDefaultTag(ref)
+	var tag string
+	switch x := ref.(type) {
+	case reference.Canonical:
+		tag = x.Digest().String()
+	case reference.NamedTagged:
+		tag = x.Tag()
+	}
+	return dockerReference{ref: ref, tag: tag}, nil
+}
+
+// NewReference returns an Docker reference for a named reference and a tag.
+// FIXME? Our "tag" is either a tag or a digest; is that a good API?
+func NewReference(named reference.Named, tag string) types.ImageReference {
+	return dockerReference{
+		ref: named,
+		tag: tag,
+	}
+}
+
+func (ref dockerReference) Transport() types.ImageTransport {
+	return Transport
+}
+
+// ConvenientString returns a string representation convenient for the user, even if perhaps syntactically a bit ambiguous.
+// For example, for Docker, this may exclude the default Docker Hub hostname or the "library" namespace.
+// This is not purely for display: parsing the string into an ImageReference must result in an equivalent reference.
+// It is perfectly acceptable for ConvenientString() and FullyExplicitString() to return the same value if the underlying references do not have format variants.
+func (ref dockerReference) ConvenientString() string {
+	// FIXME: is this ":" right?
+	return fmt.Sprintf("%s:%s", ref.ref.Name(), ref.tag)
+}
+
+// FullyExplicitString returns a string representation which is fully explicit about its semanitcs, even if a bit inconvenient to use.
+// For example, for Docker, this should include the default Docker Hub hostname or the "library" namespace.
+// This is not purely for display: parsing the string into an ImageReference must result in an equivalent reference.
+// It is highly recommended for FullyExplicitString to be designed so that all parent namespaces of a reference are initial prefixes of FullyExplicitString().
+// It is perfectly acceptable for ConvenientString() and FullyExplicitString() to return the same value if the underlying references do not have format variants.
+func (ref dockerReference) FullyExplicitString() (string, error) {
+	// FIXME: Is this the right thing to do?
```

Comment:
This originates, will replace, (and, via the interface, generalize) `policy_eval.go:fullyExpandedDockerReference`.